devito/core/gpu_openmp.py

```diff
@@ -502,13 +646,72 @@ def _specialize_iet(cls, graph, **kwargs):
         if 'openmp' not in passes:
             passes_mapper['openmp'](graph)
 
+        # Call passes
+        for i in passes:
+            try:
+                passes_mapper[i](graph)
+            except KeyError:
+                pass
+
         # Symbol definitions
-        data_manager = DeviceOpenMPDataManager(sregistry)
-        data_manager.place_ondevice(graph)
-        data_manager.place_definitions(graph)
-        data_manager.place_casts(graph)
+        DeviceOpenMPDataManager(sregistry, options).process(graph)
 
         # Initialize OpenMP environment
         initialize(graph)
 
         return graph
+
+
+# Utils
+
+def is_on_device(maybe_symbol, gpu_fit, only_writes=False):
+    """
+    True if all given Functions are allocated in the device memory, False otherwise.
+
+    Parameters
+    ----------
+    maybe_symbol : Indexed or Function or Node
+        The inspected object. May be a single Indexed or Function, or even an
+        entire piece of IET.
+    gpu_fit : list of Function
+        The Function's which are known to definitely fit in the device memory. This
+        information is given directly by the user through the compiler option
+        `gpu-fit` and is propagated down here through the various stages of lowering.
+    only_writes : boo, optional
+        Only makes sense if `maybe_symbol` is an IET. If True, ignore all Function's
+        that do not appear on the LHS of at least one Expression. Defaults to False.
+    """
+    try:
+        functions = (maybe_symbol.function,)
+    except AttributeError:
+        assert maybe_symbol.is_Node
+        iet = maybe_symbol
+        functions = set(FindSymbols().visit(iet))
+        if only_writes:
+            expressions = FindNodes(Expression).visit(iet)
+            functions &= {i.write for i in expressions}
+
+    return all(not (f.is_TimeFunction and f.save is not None and f not in gpu_fit)
```

Comment:
what about large buffered time function, like `Buffer(O(1e2))` should this be considered or is it too empirical?

Reply:
if you use Buffer, then it means that `save != None`, so they will still be captured according to this default strategy and therefore be subjected to streaming. Does it make sense?